test_util/dcos_api_session.py

```diff
@@ -30,11 +30,111 @@ def get_args_from_env():
         'public_masters': os.environ['PUBLIC_MASTER_HOSTS'].split(','),
         'slaves': os.environ['SLAVE_HOSTS'].split(','),
         'public_slaves': os.environ['PUBLIC_SLAVE_HOSTS'].split(','),
-        'default_os_user': os.getenv('DCOS_DEFAULT_OS_USER', 'root'),
-        'ca_cert_path': os.getenv('DCOS_CA_CERT_PATH', None)}
+        'default_os_user': os.getenv('DCOS_DEFAULT_OS_USER', 'root')}
+
+
+class DcosUser:
+    """A lightweight user representation for grabbing the auth info and stashing it"""
+    def __init__(self, credentials: dict):
+        self.credentials = credentials
+        self.auth_token = None
+        self.auth_cookie = None
+        self.auth = None
+
+    def authenticate(self, dcos_url: str):
+        logging.info('Attempting authentication')
+        # explicitly use a session with no user authentication for requesting auth headers
+        r = requests.post(path_join(dcos_url, '/acs/api/v1/auth/login'), json=self.credentials, auth=None)
+        r.raise_for_status()
+        logging.info('Received authentication blob: {}'.format(r.json()))
+        self.auth_token = r.json()['token']
+        self.auth_cookie = r.cookies['dcos-acs-auth-cookie']
+        logging.info('Authentication successful')
+        # Set requests auth
+        self.auth = DcosAuth(self.auth_token)
+
+
+class DcosAuth(requests.auth.AuthBase):
+    def __init__(self, auth_token: str):
+        self.auth_token = auth_token
+
+    def __call__(self, request):
+        request.headers['Authorization'] = 'token={}'.format(self.auth_token)
+        return request
+
+
+class ARNodeApiClientMixin:
```

Comment:
> That is not accurate. This is for when you are supplying node INSTEAD of host and port

What led me to my original comment was the API signature varying only with extra arg *node*

```
class ARNodeApiClientMixin:
    def api_request(self, method, path_extension, scheme=None, host=None, query=None,
                    fragment=None, port=None, node=None, **kwargs):
```

vs

```
class ApiClientSession:
    def api_request(self, method, path_extension, scheme=None, host=None, query=None,
                    fragment=None, port=None, **kwargs):
```

Agree to the arguments made in favor of Mixin here. 

Question: If we want to a "Node" specific mixin, is a good idea to change the API signature to

```
class ARNodeApiClientMixin:
    def api_request(self, method, path_extension, node=None, scheme=None, query=None,
                    fragment=None, **kwargs):
```
?

* That is, remove the host, port as a keyword arg in the API itself.
* That will eliminate the assertion errors of having all of node, host, port in the current method.

The rest of the method can stay the same.

What do you think?

(Also, irrespective of this, I think, this method can greatly benefit from keyword-only argument, like 
```
def api_request(self, method, path_extension, *, scheme=None, host=None, query=None,
                    fragment=None, port=None, node=None, **kwargs):
```
That will force providing correct keywords and as keyword args only by the caller.